Micropython Example of loading sprites

fatfingers23 opened the pull request:
* Added a flag to Presto for `sprites` if set to true, sets the pen type to `PEN_RGB332` to allow for `RGB332` sprites that work with PicoGraphics
* Added `awesome_game.py`, which is a port of the same name from the [tutfy2040](https://github.com/pimoroni/pimoroni-pico/blob/main/micropython/examples/tufty2040/awesome_game.py) example

The game uses a touch screen instead of a gamepad, so it's easier for everyone to try it; also, every time you get a treasure chest, a backlight turns on as gold.

Gadgetoid commented:
Awesome, thank you, going to run and test this!

I have a feeling that the only reason larger pen types don't support sprites is the RAM constraints. I wonder if that should be fixed now we have Pico 2 and... y'know... ALL THE PSRAM 😆

fatfingers23 commented:
It was a fun one to work on!

And it does look like the [RGB565 pen can load in sprites](https://github.com/pimoroni/pimoroni-pico/blob/4e20ac18946c4b6d51337c2a6feb90339968a031/libraries/pico_graphics/pico_graphics_pen_rgb565.cpp#L39), I may play with that some and see about changing the parameter to just be an overwritable pen type? 

I went with RGB332 cause there were sprite sheets for it, and I had no idea what magic was happening in `spritesheet-to-rgb332.py` 😅. I figured a known working state would be a good start.

fatfingers23 commented:
I did some more digging on RGB565, and at this point, it's over my head, and I'm not sure why the colors are off. But I did do some testing. 

I found the pirate themed RGB565 sprite sheets in the picosystem [repo](https://github.com/pimoroni/picosystem/tree/e9bb9571b4a3c4a8574c11df50fedb5526bfce39/assets) and set the device to RGB565 and loaded it up. Everything is a bit pink. Did move from 30fps to 60fps though. Not sure if that's because of different pens or fewer colors. Those pictures are at the bottom. 

I also found another [magic python file](https://github.com/pimoroni/picosystem/blob/main/tools/convert-image.py). I got the same results when trying to convert a spritesheet I had that was working with `RGB332`

![IMG_1418](https://github.com/user-attachments/assets/59ae7cc1-8886-4e8b-8abc-31ee9f362d46)
![IMG_1413](https://github.com/user-attachments/assets/b7dcbd1d-323f-4a25-a2d7-f8f3e284d4cb)

Gadgetoid commented:
> I had no idea what magic was happening in `spritesheet-to-rgb332.py`

Surprisingly little! I think it's probably rendered horribly cryptic by numpy operations IIRC but the principle is simple enough.

Take an 8bit per pixel image... and just chop 3, 3 and 2 bits off for red, green and blue respectively, pack 'em into a byte and repeat. RGB565 is the same but two bytes. RGB888 ... well you get the idea 😆

fatfingers23 commented:
I ended up asking a friend (Okay, maybe it was AI) and was able to get the script working to convert the sprite sheets over. Working perfectly with the RGB565 pen and even running at 60fps+

* Added the new script to a root folder called `tools` mimicking what was done in the picosystem repo. May want to double check the comments in the `image_to_data` method since it was written by AI.
* Reverted the `sprites` param that was added to `presto.py` since RGB565 sprites work fine
* I found the pngs for the sprites, so I converted them with the new script so it's no longer a pink hue
* Added a fps counter to the top right that runs during the game play loop. Not 100% sure if it's accurate, but I think it's close enough

Gadgetoid commented:
Sorry for the radio silence here, been focussed on pushing the rest of Presto over the not-quite-finished-but-good-enough line.

Would you mind giving this one last rebase, please?

fatfingers23 commented:
No problem at all @Gadgetoid ! 

I just rebased, and it should be good to go, but I will pull the uf2 down once my repo makes the build to double check it still works with the changes from the rebase

Gadgetoid commented:
Awesome, thank you!

If you want to get ahead of the latest QA/linting fixes in dev, you can also run:

```
pip install ruff
ruff check --select "F,Q,W,E,B,COM,BLE,C4,ISC,ICN,PIE,RSE,RET,SLF,ARG" --ignore "E501,E402,COM812" examples/
```

If it complains about anything that doesn't make sense to fix - unused args you really don't need to use for example - then use a `  # noqa <ERRNO>` to squash a particular error on the line.

Note: If you don't want to run the gauntlet I'll run a QA pass when I merge and rebase dev onto main anyway!

fatfingers23 commented:
Sweet! I went through and ran that and took the fixes. The only issue I ran into is the `manifest.txt`  does not seem to be picking up the `.16bpp` files to add to the firmware build, but other than that should be good to go!

Gadgetoid commented:
> The only issue I ran into is the `manifest.txt` does not seem to be picking up the `.16bpp` files

Oooh, perhaps you have found an obscure bug 😆  I'll look into that.

Thanks for adding the fixes, saves me some hassle later!

Gadgetoid commented:
Okay the `*.16bpp` problem is because there's a trailing space right after it which gets used as part of the wildcard and doesn't match any files. I should probably fix dir2uf2 to handle that more gracefully. Nice fuzz test 😆 Delete the space and you're gold!

Gadgetoid commented:
Oh also:

```
# ICON [[(-6.09, 20.0), (-7.14, 19.96), (-8.63, 19.78), (-10.29, 19.37), (-11.02, 19.11), (-12.48, 18.43), (-14.22, 17.27), (-15.52, 16.12), (-16.4, 15.14), (-17.1, 14.22), (-18.16, 12.38), (-18.65, 11.2), (-19.11, 9.61), (-19.36, 7.97), (-19.42, 6.62), (-19.38, 5.51), (-19.18, 3.92), (-18.88, 2.66), (-18.41, 1.33), (-17.67, -0.14), (-16.69, -1.58), (-15.84, -2.55), (-14.4, -3.87), (-13.45, -4.56), (-12.16, -5.31), (-10.14, -6.11), (-9.01, -6.39), (-7.41, -6.62), (-5.64, -6.67), (-4.29, -8.95), (-3.72, -9.57), (-3.13, -9.89), (-1.99, -10.1), (-0.79, -9.76), (0.58, -8.98), (1.09, -9.81), (2.19, -10.97), (3.05, -11.48), (4.52, -11.93), (5.79, -12.0), (6.89, -11.81), (7.63, -11.53), (9.64, -10.4), (7.87, -7.33), (6.31, -8.22), (5.76, -8.42), (4.89, -8.34), (4.09, -7.86), (3.64, -7.2), (5.42, -6.18), (6.24, -5.44), (6.6, -4.67), (6.71, -3.51), (6.42, -2.62), (5.2, -0.44), (5.84, 0.67), (6.61, 2.64), (6.97, 4.01), (7.21, 5.68), (7.24, 6.93), (7.07, 8.92), (6.58, 10.95), (6.05, 12.29), (5.46, 13.42), (4.79, 14.43), (3.52, 15.94), (1.73, 17.51), (0.37, 18.39), (-1.78, 19.33), (-3.76, 19.82), (-6.04, 20.0)], [(-6.09, 16.44), (-4.82, 16.37), (-3.91, 16.22), (-2.71, 15.87), (-1.8, 15.49), (-0.58, 14.77), (0.12, 14.22), (1.31, 13.05), (1.98, 12.2), (2.87, 10.64), (3.33, 9.36), (3.66, 7.41), (3.63, 5.64), (3.46, 4.6), (3.06, 3.23), (2.2, 1.46), (1.02, -0.4), (2.89, -3.6), (-1.73, -6.27), (-3.6, -3.07), (-5.56, -3.07), (-6.96, -2.99), (-8.83, -2.61), (-10.73, -1.84), (-12.3, -0.84), (-13.31, 0.04), (-14.04, 0.85), (-14.84, 2.07), (-15.34, 3.18), (-15.77, 4.78), (-15.9, 6.13), (-15.8, 8.14), (-15.36, 9.93), (-14.76, 11.29), (-14.07, 12.38), (-13.1, 13.52), (-12.35, 14.21), (-10.96, 15.18), (-9.77, 15.76), (-8.63, 16.13), (-7.05, 16.4), (-6.13, 16.44)], [(14.09, -4.89), (14.09, -8.44), (19.42, -8.44), (19.42, -4.89), (14.09, -4.89)], [(4.31, -14.67), (4.31, -20.0), (7.87, -20.0), (7.87, -14.67), (4.31, -14.67)], [(12.98, -11.07), (10.49, -13.56), (14.27, -17.33), (16.76, -14.84), (12.98, -11.07)]]
```

And

```
# NAME Awesome Game
```

fatfingers23 commented:
D'oh. Worked perfectly! Thank you so much and the icon is da bomb. 

I think it is ready to go. Thanks for all the help!

Gadgetoid commented:
No worries, it's a fantastic example and dangerously addictive 😆 absolute shoe-in. Thank you for your efforts and responsiveness here!
